Templating/VOKEmbeddedTemplateExtraction.m

```diff
@@ -0,0 +1,115 @@
+//
+//  VOKEmbeddedTemplateExtraction.m
+//  Cat2Cat
+//
+//  Created by Isaac Greenspan on 8/15/15.
+//  Copyright (c) 2015 Vokal. All rights reserved.
+//
+
+#import "VOKEmbeddedTemplateExtraction.h"
+
+#import <mach-o/getsect.h>
+#import <mach-o/ldsyms.h>
+
+#import <ZipZap.h>
+
+/**
+ *  Get the content embedded into the running mach-o executable in the __c2c_tmplt_zip section of the __TEXT segment.
+ *
+ *  @return An NSData object containing the content.
+ */
+static NSData *GetEmbeddedTemplateData(void)
+{
+    // Extract the raw bytes of the __c2c_tmplt_zip (Cat2Cat's template zip archive) section of the __TEXT segment of
+    // the currently-running mach-o executable.
+    unsigned long size;
+    void *ptr = getsectiondata(&_mh_execute_header, "__TEXT",
+                               "__c2c_tmplt_zip", &size);
+    
+    // Wrap those bytes in an NSData object.
+    return [NSData dataWithBytesNoCopy:ptr
+                                length:size
+                          freeWhenDone:NO];
+}
+
+/**
+ *  Safely (relatively securely) create a temporary directory.
+ *
+ *  @return The URL of the created temporary directory.
+ */
+static NSURL *SafelyCreateTemporaryDirectory(void)
+{
+    // Set up the directory name template.
+    NSString *tempDirectoryTemplate = [NSTemporaryDirectory() stringByAppendingPathComponent:
+                                       @"Cat2Cat-Template.XXXXXX"];  // The Xs will be replaced by random characters.
+    const char *tempDirectoryTemplateCString = [tempDirectoryTemplate fileSystemRepresentation];
+    char *tempDirectoryNameCString = (char *)malloc(strlen(tempDirectoryTemplateCString) + 1);
+    strcpy(tempDirectoryNameCString, tempDirectoryTemplateCString);
+    
+    // Have the system attempt to create a directory from the template.
+    char *result = mkdtemp(tempDirectoryNameCString);
```

Comment:
What's the advantage of using `mkdtemp` instead of just using the `NSFileManager` to create a directory which you then delete later?